modules/core/src/ocl.cpp

```diff
@@ -4715,11 +4718,14 @@ class OpenCLAllocator : public MatAllocator
                 }
 #endif
 
-                cl_int retval = 0;
-                u->data = (uchar*)clEnqueueMapBuffer(q, (cl_mem)u->handle, CL_TRUE,
-                                                     (CL_MAP_READ | CL_MAP_WRITE),
-                                                     0, u->size, 0, 0, 0, &retval);
-                if(u->data && retval == CL_SUCCESS)
+                cl_int retval = CL_SUCCESS;
+                if (CV_XADD(&(u->mapcount), 1) == 0)
```

Comment:
UMatData access is under synchronization lock (UMatDataLock) here (except `deallocate()` for some reason).